File: algorithms-data-structures/go/src/algorithms/linked-list/merge.go
package linked_list

type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	// Create a dummy node as the starting point
	dummy := &ListNode{}
	tail := dummy

	// Traverse both lists, merging them
	for list1 != nil && list2 != nil {
		if list1.Val < list2.Val {
			tail.Next = list1
			list1 = list1.Next
		} else {
			tail.Next = list2
			list2 = list2.Next
		}
		tail = tail.Next
	}

	// If one of the lists is not fully traversed, attach it to the merged list
	if list1 != nil {
		tail.Next = list1
	} else {
		tail.Next = list2
	}

	return dummy.Next
}

File: algorithms-data-structures/ts/src/algorithms/linked-list/merge.ts
export class Node {
  value: number;
  next: Node | null;

  constructor(value: number) {
    this.value = value;
    this.next = null;
  }
}

export function merge(t1: Node | null, t2: Node | null): Node | null {
  if (!t1) return t2;
  if (!t2) return t1;

  let dummy = new Node(0);
  let tail = dummy;

  while (t1 !== null && t2 !== null) {
    if (t1.value < t2.value) {
      tail.next = t1;
      t1 = t1.next;
    } else {
      tail.next = t2;
      t2 = t2.next;
    }
    tail = tail.next;
  }

  if (t1 != null) {
    tail.next = t1;
  } else {
    tail.next = t2;
  }
  return dummy.next;
}

export function printList(list: Node | null) {
  let xs = [];
  while (list !== null) {
    xs.push(list.value);
    list = list.next;
  }
  return xs;
}
